core/core-workflow-steps/src/main/java/org/eclipse/sw360/antenna/csvreader/CSVReader.java

```diff
@@ -394,11 +399,16 @@ private static String mapCPEIdToString(Artifact artifact) {
         .orElse("");
     }
 
-    private static String mapFileNameToString(Artifact artifact) {
-        return artifact.askFor(ArtifactSourceFile.class)
+    private static String getFilepathAsString(Artifact artifact) {
+        Optional<Path> path = artifact.askFor(ArtifactSourceFile.class)
                 .map(ArtifactFactWithPayload::get)
                 .map(Path::toAbsolutePath)
-                .map(Path::toString)
-                .orElse("");
+                .filter(pth -> pth.toFile().exists());
+        return path.map(Path::toString)
```

Comment:
Is there a reason why you have an obsolete path variable here? You could start with return artifact.askFor (...);

If I get you right, you should stop prior to the filter method. There are two cases for logging.
1. There is no ArtifactSourceFileFact. This is a problem, because we require to have a source file linked. This should be mentioned in output.
2. THere is an ArtifactSourceFileFact, but the the path does not reference an existing file, this is what you want to log below.

This requires that you break the optiona cascade prior to the .filter call. THere you can check whether you have a path and if you have a path to check if it exists. If both is good, you can continue by mapping the path to a string.

Reply:
> Is there a reason why you have an obsolete path variable here? You could start with return artifact.askFor (...); 

 I found that the statement itself was getting a bit to complex to be just one return statement. 

> There is no ArtifactSourceFileFact. This is a problem, because we require to have a source file linked. This should be mentioned in output.

This is just a csv reader. This function is part of _writing_ artifacts to a csv file. In our direct use case this wouldn't be a problem, it would only mean that an artifact exported from SW360 does not have a source file, which a team could fill in the updater part of the compliance tool. It is just like any other missing fact and it does not require any additional output.  

My suggestion:
```
Optional<Path> path = artifact.askFor(ArtifactSourceFile.class)
                .map(ArtifactFactWithPayload::get);
        if (path.isPresent()) {
            return path.map(Path::toAbsolutePath)
                    .filter(pth -> pth.toFile().exists())
                    .map(Path::toString)
                    .orElseGet(() -> {
                                artifact.getMainCoordinate().ifPresent(coordinate ->
                                        LOGGER.debug("The given source file for artifact {} does not exist", coordinate));
                                return "";
                            });
        } else {
            return "";
        }
```

This way we don't give a wrong warning when the fact isn't present. and give a logger debug when the fact is present but the file doesn't exist

Reply:
How about
`return artifact.askFor(ArtifactSourceFile.class)
    .map(ArtifactFactWithPayload::get)
    .map(this::getPathAsString)
    .orElse("");`

And the rest is then in this second method getPathAsString. This mixing of procedural and functional programming looks always a bit ugly.

Reply:
that is possible, however I still prefer the above method.
Since we would have to give the artifact, or at least coordinate of the artifact to the `getPathAsString` method as well for the log message. and since we give a path and not n optional would lose the easy functional programming and make it procedural. For such a small method that seems over the top

Reply:
```
private static String getFilepathAsString(Artifact artifact) {
        return artifact.askFor(ArtifactSourceFile.class)
                .map(ArtifactFactWithPayload::get)
                .map(pth -> CSVReader.getPathAsString(pth, artifact))
                .orElse("");
    }

    private static String getPathAsString(Path path, Artifact artifact) {
        File file = path.toAbsolutePath().toFile();
        if (file.exists()) {
            return file.toString();
        } else {
            artifact.getMainCoordinate().ifPresent(coordinate ->
                    LOGGER.debug("The given source file for artifact {} does not exist", coordinate));
            return "";
        }
    }
```